pkg/collect/rook.go

```diff
@@ -0,0 +1,163 @@
+)
+
+type CmdResult struct {
+	Stdout []byte
+	Stderr []byte
+}
+
+type RookCmd struct {
+	Cmd      []string
+	Args     []string
+	Filename string // file in the support-bundle
+}
+
+var RookCmds = []RookCmd{
+	RookCmd{
+		Cmd:      []string{"ceph", "status"},
+		Args:     []string{"-f", "json-pretty"},
+		Filename: "status",
+	},
+	RookCmd{
+		Cmd:      []string{"ceph", "fs", "status"},
+		Args:     []string{"-f", "json-pretty"},
+		Filename: "fs",
+	},
+	RookCmd{
+		Cmd:      []string{"ceph", "fs", "ls"},
+		Args:     []string{"-f", "json-pretty"},
+		Filename: "fs-ls",
+	},
+	RookCmd{
+		Cmd:      []string{"ceph", "osd", "status"},
+		Args:     []string{"-f", "json-pretty"},
+		Filename: "osd-status",
+	},
+	RookCmd{
+		Cmd:      []string{"ceph", "osd", "tree"},
+		Args:     []string{"-f", "json-pretty"},
+		Filename: "osd-tree",
+	},
+	RookCmd{
+		Cmd:      []string{"ceph", "osd", "pool", "ls", "detail"},
+		Args:     []string{"-f", "json-pretty"},
+		Filename: "osd-pool",
+	},
+	RookCmd{
+		Cmd:      []string{"ceph", "health", "detail"},
+		Args:     []string{"-f", "json-pretty"},
+		Filename: "health",
+	},
+	RookCmd{
+		Cmd:      []string{"ceph", "auth", "ls"},
+		Args:     []string{"-f", "json-pretty"},
+		Filename: "auth",
+	},
+}
+
+func Rook(ctx *Context) ([]byte, error) {
+	client, err := kubernetes.NewForConfig(ctx.ClientConfig)
+	if err != nil {
+		return nil, errors.New("Failed to create kubernetes clientset")
+	}
+
+	namespace := "rook-ceph"
+	selector := []string{"app=rook-ceph-operator"}
+	pods, podsErrors := listPodsInSelectors(client, namespace, selector)
+	if len(podsErrors) > 0 {
+		return nil, errors.New("Unable to find pod")
+	}
+
+	rookOutput := make(map[string][]byte)
+	for _, c := range RookCmds {
+		errCh := make(chan error, 1)
+		resultCh := make(chan CmdResult, 1)
+		go func() {
+			// TODO: find pod with 'Running' status, for now use first pod in the array
+			// Handles case where multiple pods which match select, i.e. one pod terminating
+			stdout, stderr, err := runRookCmd(ctx, client, pods[0], c.Cmd, c.Args)
+			if err != nil {
+				errCh <- err
+			} else {
+				resultCh <- CmdResult{Stdout: stdout, Stderr: stderr}
+			}
+		}()
+
+		select {
+		case <-time.After(10 * time.Second):
+			err := errors.New("Cmd timed out")
+			rookOutput[fmt.Sprintf("rook/status/%s-errors.json", c.Filename)] = []byte(err.Error())
+		case result := <-resultCh:
+			rookOutput[fmt.Sprintf("rook/status/%s.json", c.Filename)] = result.Stdout
+			if len(result.Stderr) > 0 {
+				rookOutput[fmt.Sprintf("rook/status/%s-errors.json", c.Filename)] = result.Stderr
+			}
+		case err := <-errCh:
+			rookOutput[fmt.Sprintf("rook/status/%s-errors.json", c.Filename)] = []byte(err.Error())
+		}
+	}
+
+	b, err := json.MarshalIndent(rookOutput, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+
+	return b, err
+}
+
+func runRookCmd(ctx *Context, client *kubernetes.Clientset, pod corev1.Pod, cmd, args []string) ([]byte, []byte, error) {
```

Comment:
this function feels like a copy from the run collector. what's unique about it that makes it not desirable to reuse that function?

Reply:
Yes, copyFile and getExecOutputs both use a similar function.  The same argument could have been made there as well.  There's a few minor differences though.  For one, each passes its own collector object so I couldn't just reuse those functions without reorganizing a lot of code.  Here, runRookCmd didn't require a collector function.  

On this pass I didn't want to spend a lot of time refactoring a lot of existing code.  At some point we might want to reorganize the workflow of these collectors and analyzers.  There was some rigidity in other places of code that I didn't want to spend a lot of time refactoring, such as passing functions down through a call stack as one example..  It probably would have been better to build these into Go objects.

Reply:
See the latest commit.  I moved this to become a more generic function called execPodCmd under util.go.   I'd rather change Exec and Copy later so I don't keep delaying these changes.  We can refactor those functions on a followup commit.